autograd.h

```diff
@@ -382,24 +393,118 @@ AUTOGRAD_OPTIMIZER_CLASS(SGD) {
   AUTOGRAD_KWARG(SGD, double, weight_decay, 0, 0);
   AUTOGRAD_KWARG(SGD, bool, nesterov, false, true);
   void step() override;
+  void init_state() override;
+
+  template <class Archive>
+  void serialize(Archive & ar) { 
+    ar(CEREAL_NVP(momentum_buffers_)); 
+  }
   
+ private:
+  friend class cereal::access;
+  SGD() { }
   double lr_;
-  std::unordered_map<std::string, at::Tensor> momentum_buffers;
+  std::unordered_map<std::string, at::Tensor> momentum_buffers_;
 };
 
 AUTOGRAD_OPTIMIZER_CLASS(Adam) {
  public:
+  using OptimizerImpl::OptimizerImpl;
   Adam(Container model, double lr) : Optimizer_CRTP(model), lr_(lr) { }
   AUTOGRAD_KWARG(Adam, double, beta1, 0.9, 0.9);
   AUTOGRAD_KWARG(Adam, double, beta2, 0.999, 0.999);
   AUTOGRAD_KWARG(Adam, double, weight_decay, 0, 0);
   AUTOGRAD_KWARG(Adam, double, eps, 1e-8, 1e-8);
   void step() override;
+  void init_state() override;
+
+  template <class Archive>
+  void serialize(Archive & ar) { 
+    ar(CEREAL_NVP(step_buffer_),
+       CEREAL_NVP(exp_avg_buffer_),
+       CEREAL_NVP(exp_avg_sq_buffer_)); 
+  }
   
+ private:
+  friend class cereal::access;
+  Adam() { }
   double lr_;
-  std::unordered_map<std::string, int> step_buffer;
-  std::unordered_map<std::string, at::Tensor> exp_avg_buffer;
-  std::unordered_map<std::string, at::Tensor> exp_avg_sq_buffer;
+  std::unordered_map<std::string, int> step_buffer_;
+  std::unordered_map<std::string, at::Tensor> exp_avg_buffer_;
+  std::unordered_map<std::string, at::Tensor> exp_avg_sq_buffer_;
 };
 
 }  // namespace autograd
+
+// This is super ugly and I don't know how to simplify it
```

Comment:
Random thought: maybe you can change the `AUTOGRAD_OPTIMIZER_CLASS` macro?
```cpp
#define AUTOGRAD_OPTIMIZER_CLASS(Type) \
  class Type; \
  CEREAL_REGISTER_TYPE(autograd::Type); \
  CEREAL_REGISTER_POLYMORPHIC_RELATION(autograd::OptimizerImpl, autograd::Type); \
  class Type : public autograd::Optimizer_CRTP<Type>
```

Reply:
Doesn't work actually, you have to do CEREAL_REGISTER_TYPE in a global namespace...